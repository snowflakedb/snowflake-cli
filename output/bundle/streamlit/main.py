/Users/mraba/sources/snowflake-cli.worktree/SNOW-1980772-helper-for-displaying-snowsql-set-env-vars/main.py