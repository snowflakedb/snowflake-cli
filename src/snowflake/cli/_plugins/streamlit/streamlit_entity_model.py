# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import annotations

from typing import Literal, Optional

from pydantic import Field
from snowflake.cli.api.project.schemas.entities.common import (
    Artifacts,
    EntityModelBaseWithArtifacts,
    ExternalAccessBaseModel,
    GrantBaseModel,
    ImportsBaseModel,
)
from snowflake.cli.api.project.schemas.updatable_model import (
    DiscriminatorField,
)


class StreamlitEntityModel(
    EntityModelBaseWithArtifacts,
    ExternalAccessBaseModel,
    ImportsBaseModel,
    GrantBaseModel,
):
    type: Literal["streamlit"] = DiscriminatorField()  # noqa: A003
    title: Optional[str] = Field(
        title="Human-readable title for the Streamlit dashboard", default=None
    )
    comment: Optional[str] = Field(title="Comment for the Streamlit app", default=None)
    query_warehouse: str = Field(
        title="Snowflake warehouse to host the app", default=None
    )
    main_file: Optional[str] = Field(
        title="Entrypoint file of the Streamlit app", default="streamlit_app.py"
    )
    pages_dir: Optional[str] = Field(title="Streamlit pages", default=None)
    stage: Optional[str] = Field(
        title="Stage in which the appâ€™s artifacts will be stored", default="streamlit"
    )
    # Artifacts were optional, so to avoid BCR, we need to make them optional here as well
    artifacts: Optional[Artifacts] = Field(
        title="List of paths or file source/destination pairs to add to the deploy root",
        default=None,
    )
