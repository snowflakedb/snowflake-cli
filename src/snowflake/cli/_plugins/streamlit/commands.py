# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import logging
from pathlib import Path
from typing import Dict

import click
import typer
from click import ClickException, UsageError
from snowflake.cli._plugins.object.command_aliases import (
    add_object_command_aliases,
    scope_option,
)
from snowflake.cli._plugins.streamlit.manager import StreamlitManager
from snowflake.cli._plugins.streamlit.streamlit_entity_model import (
    StreamlitEntityModel,
)
from snowflake.cli._plugins.streamlit.streamlit_project_paths import (
    StreamlitProjectPaths,
)
from snowflake.cli.api.cli_global_context import get_cli_context
from snowflake.cli.api.commands.decorators import (
    with_experimental_behaviour,
    with_project_definition,
)
from snowflake.cli.api.commands.flags import (
    ReplaceOption,
    entity_argument,
    identifier_argument,
    like_option,
)
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.constants import ObjectType
from snowflake.cli.api.exceptions import NoProjectDefinitionError
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.output.types import (
    CommandResult,
    MessageResult,
    SingleQueryResult,
)
from snowflake.cli.api.project.definition_conversion import (
    convert_project_definition_to_v2,
)

app = SnowTyperFactory(
    name="streamlit",
    help="Manages a Streamlit app in Snowflake.",
)
log = logging.getLogger(__name__)

StreamlitNameArgument = identifier_argument(
    sf_object="Streamlit app", example="my_streamlit"
)
OpenOption = typer.Option(
    False,
    "--open",
    help="Whether to open the Streamlit app in a browser.",
    is_flag=True,
)


add_object_command_aliases(
    app=app,
    object_type=ObjectType.STREAMLIT,
    name_argument=StreamlitNameArgument,
    like_option=like_option(
        help_example='`list --like "my%"` lists all streamlit apps that begin with “my”'
    ),
    scope_option=scope_option(help_example="`list --in database my_db`"),
)


@app.command(requires_connection=True)
def execute(
    name: FQN = StreamlitNameArgument,
    **options,
):
    """
    Executes a streamlit in a headless mode.
    """
    _ = StreamlitManager().execute(app_name=name)
    return MessageResult(f"Streamlit {name} executed.")


@app.command("share", requires_connection=True)
def streamlit_share(
    name: FQN = StreamlitNameArgument,
    to_role: str = typer.Argument(
        ...,
        help="Role with which to share the Streamlit app.",
        show_default=False,
    ),
    **options,
) -> CommandResult:
    """
    Shares a Streamlit app with another role.
    """
    cursor = StreamlitManager().share(streamlit_name=name, to_role=to_role)
    return SingleQueryResult(cursor)


def _default_file_callback(param_name: str):
    from click.core import ParameterSource  # type: ignore

    def _check_file_exists_if_not_default(ctx: click.Context, value):
        if (
            ctx.get_parameter_source(param_name) != ParameterSource.DEFAULT  # type: ignore
            and value
            and not Path(value).exists()
        ):
            raise ClickException(f"Provided file {value} does not exist")
        return Path(value)

    return _check_file_exists_if_not_default


@app.command("deploy", requires_connection=True)
@with_project_definition()
@with_experimental_behaviour()
def streamlit_deploy(
    replace: bool = ReplaceOption(
        help="Replace the Streamlit app if it already exists."
    ),
    entity_id: str = entity_argument("streamlit"),
    open_: bool = OpenOption,
    **options,
) -> CommandResult:
    """
    Deploys a Streamlit app defined in the project definition file (snowflake.yml). By default, the command uploads
    environment.yml and any other pages or folders, if present. If you don’t specify a stage name, the `streamlit`
    stage is used. If the specified stage does not exist, the command creates it. If multiple Streamlits are defined
    in snowflake.yml and no entity_id is provided then command will raise an error.
    """

    cli_context = get_cli_context()
    pd = cli_context.project_definition
    if not pd.meets_version_requirement("2"):
        if not pd.streamlit:
            raise NoProjectDefinitionError(
                project_type="streamlit", project_root=cli_context.project_root
            )
        pd = convert_project_definition_to_v2(cli_context.project_root, pd)

    streamlits: Dict[str, StreamlitEntityModel] = pd.get_entities_by_type(
        entity_type="streamlit"
    )

    streamlit_project_paths = StreamlitProjectPaths(cli_context.project_root)

    if not streamlits:
        raise NoProjectDefinitionError(
            project_type="streamlit", project_root=cli_context.project_root
        )

    if entity_id and entity_id not in streamlits:
        raise UsageError(f"No '{entity_id}' entity in project definition file.")

    if len(streamlits.keys()) == 1:
        entity_id = list(streamlits.keys())[0]

    if entity_id is None:
        raise UsageError(
            "Multiple Streamlit apps found. Please provide entity id for the operation."
        )

    # Get first streamlit
    streamlit: StreamlitEntityModel = streamlits[entity_id]
    url = StreamlitManager().deploy(
        streamlit=streamlit,
        streamlit_project_paths=streamlit_project_paths,
        replace=replace,
    )

    if open_:
        typer.launch(url)

    return MessageResult(f"Streamlit successfully deployed and available under {url}")


@app.command("get-url", requires_connection=True)
def get_url(
    name: FQN = StreamlitNameArgument,
    open_: bool = OpenOption,
    **options,
):
    """Returns a URL to the specified Streamlit app"""
    url = StreamlitManager().get_url(streamlit_name=name)
    if open_:
        typer.launch(url)
    return MessageResult(url)
