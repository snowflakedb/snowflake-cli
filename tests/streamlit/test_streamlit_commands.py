import shutil
from pathlib import Path
from textwrap import dedent
from unittest import mock

import pytest

from tests.streamlit.streamlit_test_class import (
    STREAMLIT_NAME,
    TYPER,
    StreamlitTestClass,
)


class TestStreamlitCommands(StreamlitTestClass):
    def test_list_streamlit(self, runner, mock_streamlit_ctx):
        self.mock_connector.return_value = mock_streamlit_ctx

        result = runner.invoke(["streamlit", "list"])

        assert result.exit_code == 0, result.output
        assert mock_streamlit_ctx.get_query() == "show streamlits like '%%'"

    @mock.patch(TYPER)
    def test_deploy_only_streamlit_file(self, mock_typer, project_directory, runner):

        with project_directory("example_streamlit") as tmp_dir:
            (tmp_dir / "environment.yml").unlink()
            shutil.rmtree(tmp_dir / "pages")
            result = runner.invoke(["streamlit", "deploy"])

        expected_query = dedent(
            f"""
                CREATE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                ROOT_LOCATION = '@streamlit/{STREAMLIT_NAME}'
                MAIN_FILE = 'streamlit_app.py'
                QUERY_WAREHOUSE = test_warehouse
                TITLE = 'My Fancy Streamlit';
                """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_called_with(expected_query)
        self.mock_create_stage.assert_called_once
        self.mock_put.assert_called_once_with(
            local_file_name=Path("streamlit_app.py"),
            stage_location="/test_streamlit/",
            overwrite=True,
            auto_compress=False,
        )

        mock_typer.launch.assert_not_called()

    @mock.patch(TYPER)
    def test_deploy_only_streamlit_file_no_stage(
        self, mock_typer, project_directory, runner
    ):
        with project_directory("example_streamlit_no_stage") as tmp_dir:
            (tmp_dir / "environment.yml").unlink()
            shutil.rmtree(tmp_dir / "pages")
            result = runner.invoke(["streamlit", "deploy"])

        expected_query = dedent(
            f"""
                CREATE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                ROOT_LOCATION = '@streamlit/{STREAMLIT_NAME}'
                MAIN_FILE = 'streamlit_app.py'
                QUERY_WAREHOUSE = test_warehouse;
                """
        ).strip()
        assert result.exit_code == 0, result.output

        self.mock_execute.assert_called_with(expected_query)
        self.mock_create_stage.assert_called_once()
        self.mock_put.assert_called_once_with(
            local_file_name=Path("streamlit_app.py"),
            stage_location="/test_streamlit/",
            overwrite=True,
            auto_compress=False,
        )
        mock_typer.launch.assert_not_called()

    def test_deploy_with_empty_pages(self, project_directory, runner):
        with project_directory("streamlit_empty_pages") as tmp_dir:
            (tmp_dir / "pages").mkdir(parents=True, exist_ok=True)
            result = runner.invoke(["streamlit", "deploy"])

        expected_query = dedent(
            f"""
                CREATE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                ROOT_LOCATION = '@streamlit/{STREAMLIT_NAME}'
                MAIN_FILE = 'streamlit_app.py'
                QUERY_WAREHOUSE = test_warehouse;
                """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_called_with(expected_query)
        self.mock_create_stage.assert_called_once()

        self._assert_that_exactly_those_files_were_put_to_stage(
            ["streamlit_app.py", "environment.yml"]
        )

    @mock.patch(TYPER)
    def test_deploy_only_streamlit_file_replace(
        self, mock_typer, project_directory, runner
    ):
        with project_directory("example_streamlit") as tmp_dir:
            (tmp_dir / "environment.yml").unlink()
            shutil.rmtree(tmp_dir / "pages")
            result = runner.invoke(["streamlit", "deploy", "--replace"])

        expected_query = dedent(
            f"""
                    CREATE OR REPLACE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                    ROOT_LOCATION = '@streamlit/{STREAMLIT_NAME}'
                    MAIN_FILE = 'streamlit_app.py'
                    QUERY_WAREHOUSE = test_warehouse
                    TITLE = 'My Fancy Streamlit';
                    """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_called_with(expected_query)
        self.mock_create_stage.assert_called_once()
        self.mock_put.assert_called_once_with(
            local_file_name=Path("streamlit_app.py"),
            stage_location="/test_streamlit/",
            overwrite=True,
            auto_compress=False,
        )
        mock_typer.launch.assert_not_called()

    @pytest.mark.parametrize(
        "project_name", ["example_streamlit_v2", "example_streamlit"]
    )
    @mock.patch(TYPER)
    def test_deploy_launch_browser(
        self, mock_typer, project_name, project_directory, runner
    ):

        with project_directory(project_name):
            result = runner.invoke(["streamlit", "deploy", "--open"])

        assert result.exit_code == 0, result.output

        mock_typer.launch.assert_called_once_with("https://foo.bar")

    @pytest.mark.parametrize(
        "project_name", ["example_streamlit_v2", "example_streamlit"]
    )
    def test_deploy_streamlit_and_environment_files(
        self, project_name, project_directory, runner, alter_snowflake_yml
    ):
        with project_directory(project_name) as tmp_dir:
            shutil.rmtree(tmp_dir / "pages")
            if project_name == "example_streamlit_v2":
                alter_snowflake_yml(
                    tmp_dir / "snowflake.yml",
                    parameter_path="entities.test_streamlit.artifacts",
                    value=["streamlit_app.py", "environment.yml"],
                )

            result = runner.invoke(["streamlit", "deploy"])

        expected_query = dedent(
            f"""
                CREATE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                ROOT_LOCATION = '@streamlit/{STREAMLIT_NAME}'
                MAIN_FILE = 'streamlit_app.py'
                QUERY_WAREHOUSE = test_warehouse
                TITLE = 'My Fancy Streamlit';
                """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_called_with(expected_query)
        self.mock_create_stage.assert_called_once()
        self._assert_that_exactly_those_files_were_put_to_stage(
            ["environment.yml", "streamlit_app.py"]
        )

    @pytest.mark.parametrize(
        "project_name", ["example_streamlit_v2", "example_streamlit"]
    )
    def test_deploy_streamlit_and_pages_files(
        self, project_name, project_directory, runner, alter_snowflake_yml
    ):
        with project_directory(project_name) as tmp_dir:
            (tmp_dir / "environment.yml").unlink()
            if project_name == "example_streamlit_v2":
                alter_snowflake_yml(
                    tmp_dir / "snowflake.yml",
                    parameter_path="entities.test_streamlit.artifacts",
                    value=["streamlit_app.py", "pages/"],
                )
            result = runner.invoke(["streamlit", "deploy"])

        expected_query = dedent(
            f"""
                    CREATE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                    ROOT_LOCATION = '@streamlit/{STREAMLIT_NAME}'
                    MAIN_FILE = 'streamlit_app.py'
                    QUERY_WAREHOUSE = test_warehouse
                    TITLE = 'My Fancy Streamlit';
                    """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_called_with(expected_query)
        self.mock_create_stage.assert_called_once()
        self._assert_that_exactly_those_files_were_put_to_stage(
            ["pages/my_page.py", "streamlit_app.py"]
        )

    @pytest.mark.parametrize(
        "project_name", ["streamlit_full_definition_v2", "streamlit_full_definition"]
    )
    def test_deploy_all_streamlit_files(self, project_name, project_directory, runner):
        with project_directory(project_name) as tmp_dir:
            result = runner.invoke(["streamlit", "deploy"])

        expected_query = dedent(
            f"""
                CREATE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                ROOT_LOCATION = '@streamlit/{STREAMLIT_NAME}'
                MAIN_FILE = 'streamlit_app.py'
                QUERY_WAREHOUSE = test_warehouse;
                """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_called_with(expected_query)
        self.mock_create_stage.assert_called_once()
        self._assert_that_exactly_those_files_were_put_to_stage(
            [
                "streamlit_app.py",
                "environment.yml",
                "pages/my_page.py",
                "utils/utils.py",
                "extra_file.py",
            ]
        )

    @pytest.mark.parametrize(
        "project_name, merge_definition",
        [
            (
                "example_streamlit_v2",
                {
                    "entities": {
                        "test_streamlit": {
                            "stage": "streamlit_stage",
                            "artifacts": [
                                "streamlit_app.py",
                                "environment.yml",
                                "pages/my_page.py",
                            ],
                        }
                    }
                },
            ),
            ("example_streamlit", {"streamlit": {"stage": "streamlit_stage"}}),
        ],
    )
    def test_deploy_put_files_on_stage(
        self, project_name, merge_definition, project_directory, runner
    ):
        with project_directory(
            project_name,
            merge_project_definition=merge_definition,
        ) as tmp_dir:
            result = runner.invoke(["streamlit", "deploy"])

        expected_query = dedent(
            f"""
                        CREATE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                        ROOT_LOCATION = '@streamlit_stage/{STREAMLIT_NAME}'
                        MAIN_FILE = 'streamlit_app.py'
                        QUERY_WAREHOUSE = test_warehouse
                        TITLE = 'My Fancy Streamlit';
                        """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_called_with(expected_query)
        self.mock_create_stage.assert_called_once()
        self._assert_that_exactly_those_files_were_put_to_stage(
            ["streamlit_app.py", "environment.yml", "pages/my_page.py"],
            streamlit_name=STREAMLIT_NAME,
        )

    @pytest.mark.parametrize(
        "project_name",
        ["example_streamlit_no_defaults", "example_streamlit_no_defaults_v2"],
    )
    def test_deploy_all_streamlit_files_not_defaults(
        self, project_name, project_directory, runner
    ):
        with project_directory(project_name) as tmp_dir:
            result = runner.invoke(["streamlit", "deploy"])

        expected_query = dedent(
            f"""
                CREATE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                ROOT_LOCATION = '@streamlit_stage/{STREAMLIT_NAME}'
                MAIN_FILE = 'main.py'
                QUERY_WAREHOUSE = streamlit_warehouse;
                """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_called_with(expected_query)
        self.mock_create_stage.assert_called_once()
        self._assert_that_exactly_those_files_were_put_to_stage(
            [
                "main.py",
                "streamlit_environment.yml",
                "streamlit_pages/first_page.py",
            ],
            STREAMLIT_NAME,
        )

    @pytest.mark.parametrize(
        "project_name", ["example_streamlit", "example_streamlit_v2"]
    )
    @pytest.mark.parametrize("enable_streamlit_versioned_stage", [True, False])
    @pytest.mark.parametrize("enable_streamlit_no_checkouts", [True, False])
    def test_deploy_streamlit_main_and_pages_files_experimental(
        self,
        os_agnostic_snapshot,
        enable_streamlit_versioned_stage,
        enable_streamlit_no_checkouts,
        project_name,
        project_directory,
        runner,
        alter_snowflake_yml,
    ):
        with (
            mock.patch(
                "snowflake.cli.api.feature_flags.FeatureFlag.ENABLE_STREAMLIT_VERSIONED_STAGE.is_enabled",
                return_value=enable_streamlit_versioned_stage,
            ),
            mock.patch(
                "snowflake.cli.api.feature_flags.FeatureFlag.ENABLE_STREAMLIT_NO_CHECKOUTS.is_enabled",
                return_value=enable_streamlit_no_checkouts,
            ),
        ):
            with project_directory(project_name) as tmp_dir:
                if project_name == "example_streamlit_v2":
                    alter_snowflake_yml(
                        tmp_dir / "snowflake.yml",
                        parameter_path="entities.test_streamlit.artifacts",
                        value=[
                            "streamlit_app.py",
                            "environment.yml",
                            "pages/my_page.py",
                        ],
                    )
                result = runner.invoke(["streamlit", "deploy", "--experimental"])

        if enable_streamlit_versioned_stage:
            post_create_command = f"ALTER STREAMLIT IDENTIFIER('{STREAMLIT_NAME}') ADD LIVE VERSION FROM LAST;"
        else:
            if enable_streamlit_no_checkouts:
                post_create_command = None
            else:
                post_create_command = (
                    f"ALTER STREAMLIT IDENTIFIER('{STREAMLIT_NAME}') CHECKOUT;"
                )

        expected_query = dedent(
            f"""
                   CREATE STREAMLIT IF NOT EXISTS IDENTIFIER('{STREAMLIT_NAME}')
                   MAIN_FILE = 'streamlit_app.py'
                   QUERY_WAREHOUSE = test_warehouse
                   TITLE = 'My Fancy Streamlit';
                   """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_any_call(expected_query)
        if post_create_command:
            self.mock_execute.assert_any_call(post_create_command)

        self._assert_that_exactly_those_files_were_put_to_stage(
            ["streamlit_app.py", "environment.yml", "pages/my_page.py"],
        )

    @pytest.mark.parametrize(
        "project_name", ["example_streamlit_no_stage", "example_streamlit_no_stage_v2"]
    )
    @pytest.mark.parametrize("enable_streamlit_versioned_stage", [True, False])
    def test_deploy_streamlit_main_and_pages_files_experimental_no_stage(
        self, enable_streamlit_versioned_stage, project_name, project_directory, runner
    ):
        with mock.patch(
            "snowflake.cli.api.feature_flags.FeatureFlag.ENABLE_STREAMLIT_VERSIONED_STAGE.is_enabled",
            return_value=enable_streamlit_versioned_stage,
        ):
            with project_directory(project_name) as tmp_dir:
                result = runner.invoke(["streamlit", "deploy", "--experimental"])

        if enable_streamlit_versioned_stage:
            post_create_command = f"ALTER STREAMLIT IDENTIFIER('{STREAMLIT_NAME}') ADD LIVE VERSION FROM LAST;"
        else:
            post_create_command = (
                f"ALTER STREAMLIT IDENTIFIER('{STREAMLIT_NAME}') CHECKOUT;"
            )

        expected_query = dedent(
            f"""
                CREATE STREAMLIT IF NOT EXISTS IDENTIFIER('{STREAMLIT_NAME}')
                MAIN_FILE = 'streamlit_app.py'
                QUERY_WAREHOUSE = test_warehouse;
                """
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_any_call(expected_query)
        if post_create_command:
            self.mock_execute.assert_any_call(post_create_command)
        self._assert_that_exactly_those_files_were_put_to_stage(
            ["streamlit_app.py", "environment.yml", "pages/my_page.py"],
        )

    @pytest.mark.parametrize(
        "project_name", ["example_streamlit", "example_streamlit_v2"]
    )
    def test_deploy_streamlit_main_and_pages_files_experimental_replace(
        self, project_name, project_directory, runner, alter_snowflake_yml
    ):

        with project_directory(project_name) as tmp_dir:
            if project_name == "example_streamlit_v2":
                alter_snowflake_yml(
                    tmp_dir / "snowflake.yml",
                    parameter_path="entities.test_streamlit.artifacts",
                    value=["streamlit_app.py", "environment.yml", "pages/"],
                )
            result = runner.invoke(
                ["streamlit", "deploy", "--experimental", "--replace"]
            )

        expected_query = dedent(
            f"""
                    CREATE OR REPLACE STREAMLIT IDENTIFIER('{STREAMLIT_NAME}')
                    MAIN_FILE = 'streamlit_app.py'
                    QUERY_WAREHOUSE = test_warehouse
                    TITLE = 'My Fancy Streamlit';
                    """
        ).strip()
        assert result.exit_code == 0, result.output
        self.mock_execute.assert_any_call(expected_query)
        self.mock_execute.assert_any_call(
            f"ALTER STREAMLIT IDENTIFIER('{STREAMLIT_NAME}') CHECKOUT;"
        )
        self._assert_that_exactly_those_files_were_put_to_stage(
            ["streamlit_app.py", "environment.yml", "pages/my_page.py"],
        )

    def test_share_streamlit(self, runner, mock_streamlit_ctx):
        self.mock_connector.return_value = mock_streamlit_ctx
        role = "other_role"

        result = runner.invoke(["streamlit", "share", STREAMLIT_NAME, role])

        assert result.exit_code == 0, result.output
        assert (
            mock_streamlit_ctx.get_query()
            == f"grant usage on streamlit IDENTIFIER('{STREAMLIT_NAME}') to role {role}"
        )

    def test_drop_streamlit(self, runner, mock_streamlit_ctx):
        self.mock_connector.return_value = mock_streamlit_ctx

        result = runner.invoke(["object", "drop", "streamlit", STREAMLIT_NAME])

        assert result.exit_code == 0, result.output
        assert (
            mock_streamlit_ctx.get_query()
            == f"drop streamlit IDENTIFIER('{STREAMLIT_NAME}')"
        )

    @mock.patch(
        "snowflake.cli._plugins.streamlit.manager.make_snowsight_url",
        return_value="https://foo.bar",
    )
    def test_get_streamlit_url(self, mock_url, runner, mock_streamlit_ctx):
        self.mock_connector.return_value = mock_streamlit_ctx
        result = runner.invoke(["streamlit", "get-url", STREAMLIT_NAME])

        assert result.exit_code == 0, result.output
        assert result.output == "https://foo.bar\n"

    @pytest.mark.parametrize(
        "command, parameters",
        [
            ("list", []),
            ("list", ["--like", "PATTERN"]),
            ("describe", ["NAME"]),
            ("drop", ["NAME"]),
        ],
    )
    @mock.patch("snowflake.cli._plugins.object.manager.ObjectManager.execute_query")
    def test_command_aliases(self, mock_execute_query, command, parameters, runner):
        result = runner.invoke(["object", command, "streamlit", *parameters])
        assert result.exit_code == 0, result.output

        result = runner.invoke(
            ["streamlit", command, *parameters], catch_exceptions=False
        )
        assert result.exit_code == 0, result.output

        queries = mock_execute_query.call_args_list
        assert len(queries) == 2
        assert queries[0] == queries[1]

    @pytest.mark.parametrize("entity_id", ["app_1", "app_2"])
    def test_selecting_streamlit_from_pdf(self, entity_id, project_directory, runner):

        with project_directory("example_streamlit_multiple_v2"):
            result = runner.invoke(["streamlit", "deploy", entity_id, "--replace"])

        expected_query = dedent(
            f"""
                CREATE OR REPLACE STREAMLIT IDENTIFIER('{entity_id}')
                ROOT_LOCATION = '@streamlit/None'
                MAIN_FILE = 'streamlit_app.py'
                QUERY_WAREHOUSE = 'streamlit';"""
        ).strip()

        assert result.exit_code == 0, result.output
        self.mock_execute.assert_any_call(expected_query)

        self._assert_that_exactly_those_files_were_put_to_stage(
            ["streamlit_app.py", f"{entity_id}.py"], streamlit_name=entity_id
        )

    def test_multiple_streamlit_raise_error_if_multiple_entities(
        self, os_agnostic_snapshot, project_directory, runner
    ):
        with project_directory("example_streamlit_multiple_v2"):
            result = runner.invoke(["streamlit", "deploy"])

        assert result.exit_code == 2, result.output
        assert result.output.strip() == os_agnostic_snapshot

    def test_deploy_streamlit_with_comment_v2(self, project_directory, runner):
        with project_directory("example_streamlit_with_comment_v2") as tmp_dir:
            result = runner.invoke(["streamlit", "deploy", "--replace"])

        expected_query = dedent(
            f"""
                CREATE OR REPLACE STREAMLIT IDENTIFIER('test_streamlit_deploy_snowcli')
                ROOT_LOCATION = '@streamlit/test_streamlit_deploy_snowcli'
                MAIN_FILE = 'streamlit_app.py'
                QUERY_WAREHOUSE = xsmall
                TITLE = 'My Streamlit App with Comment'
                COMMENT = 'This is a test comment';
                """
        ).strip()
        assert result.exit_code == 0, result.output
        self.mock_execute.assert_any_call(expected_query)
        self.mock_create_stage.assert_called_once()
        self._assert_that_exactly_those_files_were_put_to_stage(
            ["streamlit_app.py", "environment.yml", "pages/my_page.py"],
            streamlit_name="test_streamlit_deploy_snowcli",
        )

    def test_execute_streamlit(self, runner, mock_streamlit_ctx):
        self.mock_connector.return_value = mock_streamlit_ctx
        result = runner.invoke(["streamlit", "execute", STREAMLIT_NAME])

        assert result.exit_code == 0, result.output
        assert result.output == f"Streamlit {STREAMLIT_NAME} executed.\n"
        assert mock_streamlit_ctx.get_queries() == [
            "EXECUTE STREAMLIT IDENTIFIER('test_streamlit')()"
        ]
