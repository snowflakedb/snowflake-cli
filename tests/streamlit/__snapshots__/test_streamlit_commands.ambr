# serializer version: 1
# name: TestStreamlitCommands.test_multiple_streamlit_raise_error_if_multiple_entities
  '''
  Usage: root streamlit deploy [OPTIONS] [ENTITY_ID]
  Try 'root streamlit deploy --help' for help.
  +- Error ----------------------------------------------------------------------+
  | Multiple entities of type streamlit found. Please provide entity id for the  |
  | operation.                                                                   |
  +------------------------------------------------------------------------------+
  '''
# ---
